File: Algorithm/leetcode_572.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# File              : leetcode_572.py
# Author            : Andy
# Date              : 2022.08.25
# Last Modified Date: 2022.08.25
# Last Modified By  : Andy


class Solution:

    def isSubtree(self, root, subtree):

        def isSame(left, right):
            if not left and not right: return True

            if not left or not right: return False

            return left.val == right.val and isSame(
                left.left, right.left) and isSame(left.rifht, right.right)

        if not root and not subtree: return True

        if not root or not subtree: return False

        return isSame(root, subtree) or isSubtree(
            root.left, subtree) or isSubtree(root.right, subtree)

File: Algorithm/leetcode_572.go
/**
 * File              : leetcode_572.go
 * Author            : Andy
 * Date              : 2022.08.25
 * Last Modified Date: 2022.08.25
 * Last Modified By  : Andy
 */

package isSubtree

func isSame(left *TreeNode, right *TreeNode) bool {
	if left == nil && right == nil {
		return true
	}

	if left == nil || right == nil {
		return false
	}

	return left.Val == right.Val && isSame(left.Left, right.Left) && isSame(left.Right, right.Right)

}

func isSubtree(root *TreeNode, subtree *TreeNode) bool {
	if root == nil && subtree == nil {
		return true
	}

	if root == nil || subtree == nil {
		return false
	}

	return isSame(root, subtree) || isSubtree(root.Left, subtree) || isSubtree(root.Right, subtree)
}
